### PR TITLE
Remove deprecated services lead bundle - part #1

### PR DESCRIPTION
**Pull request by p365labs**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | #8137
| Issues addressed (#s or URLs) | #8138 
| BC breaks? | Y
| Deprecations? | Y

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR is trying to remove all deprecations from LeadBundle, a reference could be found is these PR #8137 and inside this issue #8138.

#### List deprecations along with the new alternative (updated so far):
1. removed LeadModel::getCurrentLead alternative *-->* ContactTracker::getContact
2. removevd LeadModel::flattenFields alternative *-->* Lead entity's getProfileFields()
3. removed LeadModel::setSystemCurrentLead alternative *-->* ContactTracker::setSystemContact
4. **[to be done in part#2]** removing LeadModel::mergeLeads alternative *-->* ContactMerger::merge

### CODE REVIEW

**Comment by p365labs** [comment]
hi @escopecz right now I have a doubt, I mean there's a problem of circular reference:

- LeadModel::mergeLeads deprecated in favor of ContactMerger::merge

The problem is ContactMerger uses LeadModel (and so injected inside the constructor).
Obviously to replace *mergeLeads* with *ContactMerger::merge* I shuld inject ContactMerger in LeadModel. So this create a **CircularReference** Exception. 
Saying that I would like to understand if you have already found how to solve it and suggest me, or if I can think by my own how to solve.

**Comment by escopecz** [comment]
I don't see a clean way how to get around it without bigger refactoring. @alanhartless do you have a plan?

**Comment by p365labs** [comment]
@escopecz Yep I agree. I'll start thinking on it but if u have suggestions I could start implement or proving something

**Comment by alanhartless** [comment]
@p365labs LeadModel::mergeLeads should be removed. Anything using LeadModel::mergeLeads should be refactored to use ContactMerger instead. That should remove the circular dependency issue.

**Comment by p365labs** [comment]
@alanhartless yep I got the point.. but the problem is the method *checkDuplicates* inside LeadModel which need to call the merge. I have to remove this method from LeadModel and move somewhere else. I was thinking about ContactDeduper. I still have found something similare to checkDuplicates. What do you think?

**Comment by escopecz** [comment]
@p365labs what is the state of this PR? We are getting closer to the beta release. It'll be last chance to do BC changes until Mautic 4.

**Comment by p365labs** [comment]
> @p365labs what is the state of this PR? We are getting closer to the beta release. It'll be last chance to do BC changes until Mautic 4.

Hi @escopecz just give me a couple of days.

**Comment by p365labs** [comment]
@alanhartless sorry for being so late :( 
So the build is not failling anymore. 

two deprecations left:
- LeadModel::isContactable -->I'm working on this
- LeadModel::mergeLeads --> after isContactable

As the PR is growing a lot, *80 files*, what if we split the LeadBundle refactoring into 2 different parts and proceed testing and merging this one; and then on the second part I'll refactor the last part? From my point of view it will take another week to close everything and as the development is really active I should rebase a lot during these days. So in order to not create other conflicts I was thinking to split this PR in 2. What do u think? @alanhartless & @escopecz 

If u don't like the approach I will continue with this PR refactoring the other deprecations.

File: app/bundles/CampaignBundle/Tests/CampaignTestAbstract.php
Comment (escopecz) [inline app/bundles/CampaignBundle/Tests/CampaignTestAbstract.php]: Let's use the shorter syntax for the newly added code if possible. It improves readability.
```suggestion
        $contactTracker = $this->createMock(ContactTracker::class);
```

File: app/bundles/EmailBundle/Model/EmailModel.php
Comment (escopecz) [inline app/bundles/EmailBundle/Model/EmailModel.php]: Let's add the namespace to the use statement at the top of the file. Reasons:
1. Shorter.
2. More readable.
3. Consistency.

File: app/bundles/EmailBundle/Model/EmailModel.php
Comment (escopecz) [inline app/bundles/EmailBundle/Model/EmailModel.php]: Shorten it here as well.

File: app/bundles/EmailBundle/MonitoredEmail/Processor/FeedbackLoop.php
Comment (escopecz) [inline app/bundles/EmailBundle/MonitoredEmail/Processor/FeedbackLoop.php]: Here again. Also on line 45.

File: app/bundles/EmailBundle/MonitoredEmail/Processor/Unsubscribe.php
Comment (escopecz) [inline app/bundles/EmailBundle/MonitoredEmail/Processor/Unsubscribe.php]: Again here and line 51

File: app/bundles/EmailBundle/Tests/MonitoredEmail/Processor/BounceTest.php
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/MonitoredEmail/Processor/BounceTest.php]:
```suggestion
        $doNotContact = $this->createMock(DoNotContact::class)
```

File: app/bundles/EmailBundle/Tests/MonitoredEmail/Processor/BounceTest.php
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/MonitoredEmail/Processor/BounceTest.php]:
```suggestion
        $doNotContact = $this->createMock(DoNotContact::class);
```

File: app/bundles/EmailBundle/Tests/MonitoredEmail/Processor/ReplyTest.php
Comment (escopecz) [inline app/bundles/EmailBundle/Tests/MonitoredEmail/Processor/ReplyTest.php]:
```suggestion
        $contactTracker = $this->createMock(ContactTracker::class);
```

File: app/bundles/FormBundle/Tests/Model/DeleteFormTest.php
Comment (escopecz) [inline app/bundles/FormBundle/Tests/Model/DeleteFormTest.php]:
```suggestion
            ->createMock(ContactTracker::class);
```

File: app/bundles/LeadBundle/Controller/AjaxController.php
Comment (escopecz) [inline app/bundles/LeadBundle/Controller/AjaxController.php]:
```suggestion
            /** @var DoNotContact $doNotContact */
```

File: app/bundles/LeadBundle/Controller/AjaxController.php
Comment (escopecz) [inline app/bundles/LeadBundle/Controller/AjaxController.php]:
```suggestion
            /** @var DoNotContact $doNotContact */
```

File: app/bundles/LeadBundle/Controller/Api/LeadApiController.php
Comment (escopecz) [inline app/bundles/LeadBundle/Controller/Api/LeadApiController.php]:
```suggestion
                /** @var DoNotContact $doNotContact */
```
Plus lines 414 and 432

File: app/bundles/LeadBundle/Event/DoNotContactAddEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DoNotContactAddEvent.php]:
```suggestion
    public function __construct(Lead $lead, string $channel, string $comments = '', string $reason = DNC::BOUNCED, bool $persist = true, bool $checkCurrentStatus = true, bool $override = true)
```
Since we have minimal PHP version of 7.2 for M3, let's use all goodies we can in the newly added classes and methods.

File: app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php]:
```suggestion
    public function __construct(Lead $lead, string $channel, bool $persist = true)
```

File: app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php
Comment (escopecz) [inline app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php]: All new files are missing the file comments.

File: app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php
Comment (escopecz) [inline app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php]:
```suggestion
    public static function getSubscribedEvents(): array
```

File: app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php
Comment (escopecz) [inline app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php]:
```suggestion
    public function removeDncForLead(DoNotContactRemoveEvent $doNotContactRemoveEvent): void
```

File: app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php
Comment (escopecz) [inline app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php]:
```suggestion
    public function addDncForLead(DoNotContactAddEvent $doNotContactAddEvent): void
```

File: app/bundles/LeadBundle/Tests/Controller/Api/LeadApiControllerFunctionalTest.php
Comment (escopecz) [inline app/bundles/LeadBundle/Tests/Controller/Api/LeadApiControllerFunctionalTest.php]: Is this change intentional?

File: app/bundles/NotificationBundle/EventListener/CampaignSubscriber.php
Comment (escopecz) [inline app/bundles/NotificationBundle/EventListener/CampaignSubscriber.php]:
```suggestion
        DoNotContact $doNotContact
```
+ line 52

File: app/bundles/PageBundle/Tests/PageTestAbstract.php
Comment (escopecz) [inline app/bundles/PageBundle/Tests/PageTestAbstract.php]:
```suggestion
            ->createMock(ContactTracker::class);
```

File: app/bundles/PluginBundle/Integration/AbstractIntegration.php
Comment (escopecz) [inline app/bundles/PluginBundle/Integration/AbstractIntegration.php]:
```suggestion
        DoNotContact $doNotContact
```
+ line 196

File: app/bundles/PointBundle/Model/PointModel.php
Comment (escopecz) [inline app/bundles/PointBundle/Model/PointModel.php]: Any reason why not `private`?

File: app/bundles/PointBundle/Model/TriggerModel.php
Comment (escopecz) [inline app/bundles/PointBundle/Model/TriggerModel.php]: Any reason why not private?

**Review by escopecz — CHANGES_REQUESTED** [review]
Nice work! I found some syntax sugar that could improve the readability. Please take a look.

I agree that 2 shorter PRs are better than 1 big one. Let's merge this one.

File: app/bundles/PointBundle/Model/PointModel.php
Comment (p365labs) [reply app/bundles/PointBundle/Model/PointModel.php]: not specific one. The others properties are protected so I take them protecte.

File: app/bundles/PointBundle/Model/TriggerModel.php
Comment (p365labs) [reply app/bundles/PointBundle/Model/TriggerModel.php]: same as above

**Comment by p365labs** [comment]
@escopecz  & @alanhartless sorry the last tests have failed as there was a commit I left but it was for part #2 of this PR

**Comment by p365labs** [comment]
done.

File: app/bundles/PointBundle/Model/PointModel.php
Comment (escopecz) [reply app/bundles/PointBundle/Model/PointModel.php]: Could you please make the new ones private?

**Comment by escopecz** [comment]
@p365labs thanks! There are several more comments if you click on "Load More" or view it in the "Files changed" tab.

File: app/bundles/LeadBundle/Tests/Controller/Api/LeadApiControllerFunctionalTest.php
Comment (p365labs) [reply app/bundles/LeadBundle/Tests/Controller/Api/LeadApiControllerFunctionalTest.php]: it's a typo :( sorry

File: app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php
Comment (p365labs) [reply app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php]: done

File: app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php
Comment (escopecz) [inline app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php]: The newly added files should have the current year here. Do not change it for existing files you modify. Just the newly added.

File: app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php]: I missed this one
```suggestion
    public function __construct(Lead $lead, string $channel, bool $persist = true)
```

File: app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php]:
```suggestion
    public function getLead(): Lead
```

File: app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php]:
```suggestion
    public function getChannel(): string
```

File: app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php]:
```suggestion
    public function getPersist(): bool
```

File: app/bundles/LeadBundle/Event/DoNotContactAddEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DoNotContactAddEvent.php]:
```suggestion
    public function getLead(): Lead
```

File: app/bundles/LeadBundle/Event/DoNotContactAddEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DoNotContactAddEvent.php]:
```suggestion
    public function getChannel(): string
```

File: app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php
Comment (escopecz) [inline app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php]: This comment doesn't seem to be accurate.

File: app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php
Comment (escopecz) [inline app/bundles/LeadBundle/Event/DoNotContactRemoveEvent.php]: I don't see a reason why this class should not be final. It doesn't have to be mocked as it does not use any services. What do you think?
```suggestion
final class DoNotContactRemoveEvent extends Event
```

File: app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php
Comment (escopecz) [inline app/bundles/LeadBundle/EventListener/DoNotContactSubscriber.php]: I can't find a reason why this class should be extendable. Subscribers aren't used as dependencies of other services so there is no need to mock them.
```suggestion
final class DoNotContactSubscriber implements EventSubscriberInterface
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Amazing, thanks for fast corrections! I found a few more. Last ones, I promise :)

**Comment by p365labs** [comment]
> Amazing, thanks for fast corrections! I found a few more. Last ones, I promise :)

done! any other thing for this PR?

**Comment by escopecz** [comment]
Just a test. Could you help the tester to describe what areas of Mautic must be tested?

**Review by hluchas — CHANGES_REQUESTED** [review]
👍 Good work. Thank you.
I left some comments.

**Comment by hluchas** [comment]
Could you please rebase 3.x branch. I'm getting errors using Mautic UI which are fixed.  and there are some conflicts.
TY!

**Comment by hluchas** [comment]
@p365labs Are you able to fix this in next couple of days? I can take it and do the rest ;-)

**Comment by p365labs** [comment]
@hluchas yes I think so

File: app/bundles/FormBundle/Tests/FormTestAbstract.php
Comment (p365labs) [reply app/bundles/FormBundle/Tests/FormTestAbstract.php]: done

File: app/bundles/FormBundle/Tests/FormTestAbstract.php
Comment (p365labs) [reply app/bundles/FormBundle/Tests/FormTestAbstract.php]: done

File: app/bundles/PageBundle/Tests/PageTestAbstract.php
Comment (p365labs) [reply app/bundles/PageBundle/Tests/PageTestAbstract.php]: done

**Comment by p365labs** [comment]
@hluchas done! let me know if there is something else to fix

File: app/bundles/LeadBundle/Model/LeadModel.php
Comment (p365labs) [reply app/bundles/LeadBundle/Model/LeadModel.php]: done

**Review by hluchas — APPROVED** [review]
It is ok for me 👍 
Could you please take a look if you are OK with this and merge it? @escopecz 

Huge and good work! Thank you @p365labs !!!

**Comment by escopecz** [comment]
I would like to be clear what have we tested before we merge this PR as I [touched before](https://github.com/mautic/mautic/pull/8335#issuecomment-590964083). This PR makes changes in many several places. @p365labs could you please make it clear what needs to be tested?

**Comment by p365labs** [comment]
@escopecz not sure where every one of my updates is involved. I just replaced the deprecated code with new strategies.

**Comment by hluchas** [comment]
> @escopecz not sure where every one of my updates is involved. I just replaced the deprecated code with new strategies.

I did some acceptance testing by hand and it seems to be working well, but I'm not sure about all consequences too.

**Comment by escopecz** [comment]
It can be deduced from the changed code. @hluchas what specifically were you testing? From the changes I think we have to test:
- [ ] Asset download tracks contact that downloaded the asset.
- [ ] Campaign processing works correctly
- [ ] Dynamic web content works correctly and tracks the contact
- [ ] Email tracking works correctly
- [ ] Sending emails from point bundle works
- [ ] Setting DNC as a contact works from the unsubscribe link
- [ ] Monitored inbox works
- [ ] Form submissions work and are connected to the contact
- [ ] Toggling preferred contact channel works
- [ ] Removing bounce status (not sure how to do that in the UI)
- [ ] Removing/Adding DNC via API works. Doing single and batch requests. (Running the API library tests will do)
- [ ] Toggling DNC from the UI. Batch and single.
- [ ] Page hit tracking works for the tracking script and for landing pages
- [ ] Gated video works (https://docs.mautic.org/en/gated-video)
- [ ] Focus Items work correctly for contacts
- [ ] Pipedrive, SugarCRM and other CRM plugins
- [ ] Social plugins

It's quite a long list.

**Comment by hluchas** [comment]
After discussion we will merge this after first M3 version is released. There is no time to do comprehensive testing for now

**Comment by alanhartless** [comment]
This can and should still go into the final release of M3. @p365labs I see this is "part 1." What is planned for "part 2" and is it absolutely critical? Since this has BC breaks, we'd have to get it into M3 or wait till M4. But, we can't hold up M3 for much longer. Due to time restraints, we've unfortunately had to defer a lot of refactoring of deprecated Mautic business logic code till later.

**Comment by alanhartless** [comment]
Oh sorry - I see your "part 2" note in the description. Thoughts on timing for that?

**Comment by p365labs** [comment]
@alanhartless & @hluchas 
The part that last is related to remove these 2 deprecated methods

LeadModel::isContactable
LeadModel::mergeLeads

I think I'll have a first draft for a PR for this weekend.

Considering it's a PR of 80files I won't keep it open too much, but these are my two cents

**Review by dongilbert — APPROVED** [review]
+1 thanks @p365labs 

Cursory checks looked good. Tested many aspects of the system and they seemed working as expected.

**Comment by marioschubert** [comment]
I am not sure, if this belongs here. I see following Error, when I try to edit the preferences of a contact (3.0.1):

`[2020-07-03 21:45:26] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\UndefinedMethodException: "Attempted to call an undefined method named "removeDncForLead" of class "Mautic\LeadBundle\Model\LeadModel"." at /var/www/html/app/bundles/LeadBundle/Controller/FrequencyRuleTrait.php line 180 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\UndefinedMethodException(code: 0): Attempted to call an undefined method named \"removeDncForLead\" of class \"Mautic\\LeadBundle\\Model\\LeadModel\". at /var/www/html/app/bundles/LeadBundle/Controller/FrequencyRuleTrait.php:180)"} []`

When I click on "Re-enable Contact":

`[2020-07-03 23:57:41] mautic.CRITICAL: Uncaught PHP Exception InvalidArgumentException: "mautic.lead.model.dnc is not a valid model key." at /var/www/html/app/bundles/CoreBundle/Factory/ModelFactory.php line 41 {"exception":"[object] (InvalidArgumentException(code: 0): mautic.lead.model.dnc is not a valid model key. at /var/www/html/app/bundles/CoreBundle/Factory/ModelFactory.php:41)"} []`

**Comment by therightprofile** [comment]
> I am not sure, if this belongs here. I see following Error, when I try to edit the preferences of a contact (3.0.1):
> 
> `[2020-07-03 21:45:26] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\UndefinedMethodException: "Attempted to call an undefined method named "removeDncForLead" of class "Mautic\LeadBundle\Model\LeadModel"." at /var/www/html/app/bundles/LeadBundle/Controller/FrequencyRuleTrait.php line 180 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\UndefinedMethodException(code: 0): Attempted to call an undefined method named \"removeDncForLead\" of class \"Mautic\\LeadBundle\\Model\\LeadModel\". at /var/www/html/app/bundles/LeadBundle/Controller/FrequencyRuleTrait.php:180)"} []`
> 
> When I click on "Re-enable Contact":
> 
> `[2020-07-03 23:57:41] mautic.CRITICAL: Uncaught PHP Exception InvalidArgumentException: "mautic.lead.model.dnc is not a valid model key." at /var/www/html/app/bundles/CoreBundle/Factory/ModelFactory.php line 41 {"exception":"[object] (InvalidArgumentException(code: 0): mautic.lead.model.dnc is not a valid model key. at /var/www/html/app/bundles/CoreBundle/Factory/ModelFactory.php:41)"} []`

Same here. Did you find a solution?

**Pull request merged**
